### PR TITLE
Feature/chat locker 30

### PR DESCRIPTION
**Pull request by douglasrapp**
Ok, this one's a little big. 3 main changes:

1) Modified the 'canAccessResource' method of the 'AccessPermission' class, as well as the 'canAccessResource' Meteor method to return, in addition to pass/fail, a list of the users that cannot access, as well as the permission ids that cause them to fail.

2) Modified the 'directMessageFlex', 'privateMessageFlex', 'securityLabels', and 'userSearch' templates to obtain this information and use it to determine what items in the left pane need to be styled differently.

3) Modified the 'updateRoom' and 'createPrivateGroup' Meteor methods so allow a permission change in a private group to occur even if an existing (or specified) user does not have access. In those cases, the associated users are kicked/excluded. Also, if a user tries to add a user that cannot access the specified labels, that action is simply ignored.

A few things to point out:

1) In some cases, adding a CSS class to an element (whether in the HTML template via a helper function or via the JQuery.addClass function) still did not result in the element being styled appropriately. The style was getting overridden by other applicable CSS classes, and I was unable to figure it out.

Additionally, in other cases, it wasn't possible to use helpers in the HTML template for the Chosen stuff because it generates the applicable HTML on the fly.

So, to be consistent across all of the changes to style the labels and whatnot, I used the JQuery.css(<option>) and JQuery.removeAttr('style') functions to set and revert styling changes, respectively. I would have greatly preferred to just use a CSS class, so as to centralize the configuration and make it easier to change how the 'problem labels' appear. If anyone wants to take a look at that and see if they can figure it out, be my guest.

2) With the 'securityLabels' template, I was able to use ReactiveVars passed in the data context to communicate the needed user selection information and let the label styling be reactive to changes. However, with the 'userSearch' template (used with the meteor-autocomplete package when selecting users), there was no way (that I could see) to modify the passed data context. I needed to get a hold of the currently-selected labels so as to be able to make checks for if the users in the drop-down dialog could access them. So, the only way I could think to do that was using a Session variable.

3) When applying the logic for updating rooms and creating private groups, I'm pretty sure I covered all the edge cases. But please try to break it - e.g., try adding users to rooms they can't access, adding users along with labels they can't access, adding and removing at the same time, etc.

### CODE REVIEW

**Comment by bbrockman** [comment]
For the css overriding issue.  I may be able to help.  I had to find out why the one for #15 wasn't being styled the way I wanted it to.

**Comment by douglasrapp** [comment]
To reproduce the CSS being overridden issue, take a look at the "client/views/app/sideNav/securityLabels.js" file.

On lines 90 and 94, the JQuery .css() and .removeAttr() functions are being called to style and revert the style for the security labels (but not the users - those are handled elsewhere). Change those calls to "addClass('warning-user-label')" and "removeClass('warning-user-label')", respectively.

Then, when using a tool to examine the HTML elements in the browser, you can see that the class definitely gets applied, but the "color" directive specified in that class is far down the priority list and gets overridden by some of the built-in CSS classes for the "Chosen" plugin.

File: client/views/app/sideNav/directMessagesFlex.coffee [hunk 1]
```diff
@@ -36,6 +37,12 @@ Template.directMessagesFlex.helpers
 	selectedUser: ->
 		return Template.instance().selectedUser.get()
 
+	warning: ->
+		if Template.instance().warning.get()
+			return 'Due to security label access conflicts, you will not be able to proceed with the currently-selected labels.'
```
Comment (bbrockman) [inline client/views/app/sideNav/directMessagesFlex.coffee]: Is there a reason there is a dash after currently?

Comment (douglasrapp) [inline client/views/app/sideNav/directMessagesFlex.coffee]: Ah, that is incorrect.. I got too hyphen-happy ;). According to Rule 3 here: (http://www.grammarbook.com/punctuation/hyphens.asp), the hyphen shouldn't be there for adverbs ending in -ly. I think there might be other instances of this - I'll fix 'em.

Comment (douglasrapp) [inline client/views/app/sideNav/directMessagesFlex.coffee]: Fixed (there were a bunch of comments too :( )

**Comment by bbrockman** [comment]
One issue when relabeling.  If you select a red sap or sci, you get the message.  If you remove the red selected sap, the values in the dropdown become grey.  If you close the dropdown and reopen it, they are red (like they are supposed to be).

**Comment by bbrockman** [comment]
Style needs to be updated to be more clear, but we can do that on another sprint.

**Comment by bbrockman** [comment]
:+1:

**Pull request merged**